src/schema/posts.ts

```diff
@@ -1327,6 +1366,76 @@ export const resolvers: IResolvers<any, Context> = {
       }
       return { _: true };
     },
+    downvote: async (
+      source,
+      { id }: { id: string },
+      ctx: Context,
+    ): Promise<GQLEmptyResponse> => {
+      try {
+        const post = await ctx.con.getRepository(Post).findOneByOrFail({ id });
+        await ensureSourcePermissions(ctx, post.sourceId);
+        await ctx.con.transaction(async (entityManager) => {
+          await entityManager.getRepository(Downvote).insert({
+            postId: id,
+            userId: ctx.userId,
+          });
+
+          await saveHiddenPost(entityManager.connection, {
+            userId: ctx.userId,
+            postId: id,
+          });
+
+          const upvote = entityManager.getRepository(Upvote).findOneBy({
+            postId: id,
+            userId: ctx.userId,
+          });
+
+          if (upvote) {
+            await entityManager.getRepository(Upvote).delete({
+              postId: id,
+              userId: ctx.userId,
+            });
+          }
+        });
+      } catch (err) {
+        // Foreign key violation
+        if (err?.code === TypeOrmError.FOREIGN_KEY) {
+          throw new NotFoundError('Post or user not found');
+        }
+        // Unique violation
+        if (err?.code !== TypeOrmError.DUPLICATE_ENTRY) {
+          throw err;
+        }
+      }
+      return { _: true };
+    },
+    cancelDownvote: async (
+      source,
+      { id }: { id: string },
+      ctx: Context,
+    ): Promise<GQLEmptyResponse> => {
+      await ctx.con.transaction(async (entityManager) => {
+        const downvote = await entityManager.getRepository(Downvote).findOneBy({
+          postId: id,
+          userId: ctx.userId,
+        });
+
+        if (downvote) {
+          await entityManager.getRepository(Downvote).delete({
+            postId: id,
+            userId: ctx.userId,
+          });
+
+          await entityManager
+            .getRepository(HiddenPost)
+            .delete({ postId: id, userId: ctx.userId });
+
```

Comment:
Let's turn it into a function. We have this logic twice here already

Reply:
moved to `revertPostDownvote` 👍